### ISSUE TITLE
Game become slow in loading after replace a files

### ISSUE THREAD

**Issue by rna98mod** [nesrak1/UABEA#222]
i tried to replace mesh file in this game, the game files is in .unity3d format. I managed to uncompressed the data.unity3d file, then export the files i need (resource.assets). After importing/replacing the files i wanted, i import the files (resource.assets) back to the uncompressed data.unity3d. Modding process have no issue and no error, but when the files applied to the game, it become slow when loading a files and need about 1-2 min to just load another stuff (e.g change and choosing car in garage menu), but it works, the game read the files, just become so slow.

original
![Screenshot_2023-02-05-00-17-56-435_com fengiiley frlegends](https://user-images.githubusercontent.com/72114467/216780790-2772f3e5-b254-4260-876d-aeab5558e914.png)

modified
![Screenshot_2023-02-04-23-26-04-216_com fengiiley frlegends](https://user-images.githubusercontent.com/72114467/216780808-cff81683-2598-4555-99f1-792e4ca8b278.png)

this only happen when using the modified files. When i tried using original files (compressed and not compressed) the game works flawlessly. 

the game uses Unity 2020.3.43f1c1, platform Android.

There's some difference when i compare the original uncompressed data.unity3d with the one that i modified
![image](https://user-images.githubusercontent.com/72114467/216780322-02cf5e96-879f-4b66-9033-94569e9e6208.png)

original uncompressed data.unity3d size = 342.032 kb
modified uncompressed data.unity3d size = 342.003 kb

**Comment by nesrak1** [comment]
> Modding process have no issue and no error, but when the files applied to the game, it become slow when loading a files and need about 1-2 min to just load another stuff (e.g change and choosing car in garage menu), but it works, the game read the files, just become so slow.

Is the game lagging all the time or just taking long to load certain screens? I haven't had anyone report that, although if you change a Texture2D, I still don't support mipmaps yet, so you could have poorer performance all the time (although that game looks to have simple graphics so I'm not sure). I can't really tell by your screenshots what you want me to see, nor do I know what you changed, so it's hard to help you diagnose the problem.

> There's some difference when i compare the original uncompressed data.unity3d with the one that i modified

The original uncompressed version is showing a file where all of the LZ4 blocks converted into multiple uncompressed blocks (what UABEA does during decompression.) The modified bundle shoes a file where there's a single uncompressed block. There's no real difference on how it behaves other than how it looks in the hex editor and maybe a slight performance hit in the blocked version.

**Comment by rna98mod** [comment]
> Is the game lagging all the time or just taking long to load certain screens? I haven't had anyone report that, although if you change a Texture2D, I still don't support mipmaps yet, so you could have poorer performance all the time (although that game looks to have simple graphics so I'm not sure). I can't really tell by your screenshots what you want me to see, nor do I know what you changed, so it's hard to help you diagnose the problem.

the game is not lagging (drop fps), just taking long time to load every scene, not just scene but also when loading a gameobject, i only try to change a Mesh asset in this case.

**Comment by nesrak1** [comment]
I haven't heard of this one, but since you said the non-compressed file worked fine but only loaded slow when rewritten, it could be that the engine tries to load the whole file at once instead of the individual blocks. Maybe if you recompress the modified bundle to LZ4, it will speed back up? That would put the block chunks back.

But again, I've not heard of it taking significantly longer to load in game, so I'm not entirely sure.

**Comment by rna98mod** [comment]
I have tried to recompress the modified bundle to LZ4, and the result still the same.

I tried to re-import the extracted file (without modify it) to the bundle, recompress, and apply it to the game, and the result is the same, took long time to load a scene. 

Here',s the step i did :
1. Open data.unity3d
2. Extract it to file/memory
3. Export a file from the bundle (in this case, i extract resource.assets from the data.unity3d bundle)
4. Import the extracted file back to bundle
5. Save the bundle into different file name
6. Open the new saved bundle file
7. Compress it with LZ4
8. Copy and apply the compressed bundle to the game

With this step, although i didn't change any mesh or any assets, the game would still take long time when loading.

**Comment by nesrak1** [comment]
You said the bundle uncompressed is totally fine. I want to see if the game is doing something extra if the bundle's data CRC doesn't match (unaffected by decompression). In UABEA, look for any asset with a recognizable name. Make sure that whatever name it is does not appear elsewhere. Next, in HxD, find the string (double check there aren't multiple matches for the string) and overwrite one of the characters with anything you want (don't insert). If you do this, the CRC should change but since name fields are largely ignored, you shouldn't see any difference in-game. If the game _still_ takes too long to load, then _any_ kind of modification, no matter what tool, will always have this loading screen. In that case, it may be worth looking into the game's code to figure out what it's doing extra (redownloading the bundle from the internet, for example)

**Comment by Swagsy** [comment]
Bump, had this happen to me again with another android game

**Comment by Swagsy** [comment]
https://github.com/SeriousCache/UABE/issues/548
There's something to it, I couldn't figure out what it is

**Comment by nesrak1** [comment]
I still don't have a solution to the problem. Anyone is welcome to investigate and play around with it. I wrote something up here but no one has followed it yet to my knowledge: https://github.com/nesrak1/UABEA/issues/222#issuecomment-1433341657

**Comment by Swagsy** [comment]
Ah yeah, will test it out myself soon then, thanks!

**Comment by Swagsy** [comment]
Update: In my case, this was related to the APK Signer I was using after putting everything back together after modification. After switching to another APK signer, it improved massively

**Comment by AeroForta** [comment]
> Update: In my case, this was related to the APK Signer I was using after putting everything back together after modification. After switching to another APK signer, it improved massively

What APK Signer do you use?

**Comment by Controllerdestiny** [comment]
> 你说未压缩的束完全没问题。我想看看如果捆绑包的数据CRC不匹配（不受解压影响），游戏是否会做额外的处理。在UABEA中，寻找任何有名气的资产。确保无论哪个名字都不在其他地方出现。接下来，在《HxD》中，使用未修改、未压缩的捆绑，找到字符串（确认该字符串没有多个匹配），然后用你想要的格式覆盖其中一个字符（不要插入）。如果你这样做，CRC应该会变化，但由于名字字段大多被忽略，游戏中你应该看不到任何差别。如果游戏_加载时间仍然_太长，那么无论用什么工具，_任何_修改都会有加载画面。在这种情况下，或许值得研究一下游戏代码，弄清楚它在做什么额外的作（比如从网上重新下载捆绑包）



> You said the bundle uncompressed is totally fine. I want to see if the game is doing something extra if the bundle's data CRC doesn't match (unaffected by decompression). In UABEA, look for any asset with a recognizable name. Make sure that whatever name it is does not appear elsewhere. Next, in HxD with the unmodified, uncompressed bundle, find the string (double check there aren't multiple matches for the string) and overwrite one of the characters with anything you want (don't insert). If you do this, the CRC should change but since name fields are largely ignored, you shouldn't see any difference in-game. If the game _still_ takes too long to load, then _any_ kind of modification, no matter what tool, will always have this loading screen. In that case, it may be worth looking into the game's code to figure out what it's doing extra (redownloading the bundle from the internet, for example)

I don't believe this is a UABEA issue. Yesterday I encountered a similar situation and ultimately determined the issue stemmed from incorrectly adding data.unity3d to the installer package. Since data.unity3d itself is already compressed, including it compressed within the installer forces the program to perform unnecessary decompression steps. Therefore, data.unity3d should be added to the installer package in a read-only storage format. The fact that adding an uncompressed data.unity3d to the installer package worked as described confirms this approach.
